src/main/java/fi/aalto/cs/intellij/actions/RequiredPluginsCheckerAction.java

```diff
@@ -0,0 +1,211 @@
+package fi.aalto.cs.intellij.actions;
+
+import static java.util.stream.Collectors.toMap;
+
+import com.intellij.ide.plugins.IdeaPluginDescriptor;
+import com.intellij.ide.plugins.PluginManager;
+import com.intellij.ide.plugins.PluginManagerConfigurable;
+import com.intellij.ide.plugins.RepositoryHelper;
+import com.intellij.ide.plugins.newui.BgProgressIndicator;
+import com.intellij.notification.Notification;
+import com.intellij.notification.NotificationAction;
+import com.intellij.notification.NotificationType;
+import com.intellij.notification.Notifications;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.extensions.PluginId;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.startup.StartupActivity;
+import com.intellij.openapi.updateSettings.impl.PluginDownloader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.StringJoiner;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+//TODO THINK HOW TO TEST & REFACTOR THIS
+public class RequiredPluginsCheckerAction implements StartupActivity {
+
+  private static final Logger logger = LoggerFactory.getLogger(RequiredPluginsCheckerAction.class);
+
+  private static final Map<String, String> requiredPluginNames = new HashMap<>();
+  private static Map<String, String> missingOrDisabledPluginNames = new HashMap<>();
+  private static List<IdeaPluginDescriptor> missingOrDisabledIdeaPluginDescriptors;
+  private static List<IdeaPluginDescriptor> availableIdeaPluginDescriptors;
+
+  /**
+   * The plugin startup activity.
+   */
+  @Override
+  public void runActivity(@NotNull Project project) {
+    populateRequiredPluginNamesMap();
+    filterMissingOrDisabledPluginNames();
+    createListOfMissingOrDisabledPluginDescriptors();
+    checkPluginsStatusAndNotify();
```

Comment:
**methods style:** there are various approached, but in general:
* can be void, take no parameters, operate on class fields
* can be normal functions (smth in, something out)

Reply:
in the IJ itself, many things are using the first approach, especially when it comes to building the UI (Swing "likes" voids): https://github.com/JetBrains/intellij-community/blob/21dbcdb8b9f07b950ad8440cd217f7fc8a920366/platform/core-impl/src/com/intellij/ide/plugins/PluginManagerCore.java
but also in other places quite a lot

on the other hand, testing of void methods is kinda impossible, but if they are private as well, then not explicitly required

Reply:
"normal functions (smth in, something out)" is my personal favourite of these two (as I am always a bit scared of objects whose state changes a lot :D) but your implementation is very clear and nice so there is no strong reason to "fix" it now.
I agree that "normal functions" are usually easier to write unit tests on but I don't see how that would be impossible for void functions either (void function should cause certain side-effects, in this case change the object's state, so in a unit test we check that these changes have happened).
Also, I am not sure if a method being `private` is a good enough reason to not test it.

Reply:
accessibility of privates is usually difficult, as the tests are residing elsewhere. but there is no reason making util methods public just for the sake of testings

Reply:
If we wanted to unit test those methods, we could make them package-private (i.e. remove the `private` modifier).

Another question is, whether methods that are only meant to be used internally in a class should be unit tested at all. But in this case, I'd say it's much easier to test these four methods separately (plus some basic tests for the entirety) than to write a lot of different tests to the public interface trying to cover all the corner cases. But this is something we don't have to resolve now.

Reply:
another potential candidate, meet JB's own [ApiStatus.Internal](https://github.com/JetBrains/java-annotations/blob/450e2e8e109680eee86faaa80520d0819b1821fa/common/src/main/java/org/jetbrains/annotations/ApiStatus.java#L51)